### PR TITLE
Add CountVectorizer

### PR DESCRIPTION
**Pull request by ksew1**
Hi I added `CountVectorizer` as it would be useful in pipeline with Naive Bayes algorithms that we already implemented (for example in spam detection)

### CODE REVIEW

File: lib/scholar/feature_extraction/count_vectorizer.ex [hunk 1]
```diff
@@ -0,0 +1,147 @@
+defmodule Scholar.FeatureExtraction.CountVectorizer do
+  @moduledoc """
+    A `CountVectorizer` converts a collection of text documents to a matrix of token counts.
```
Comment (josevalim) [inline lib/scholar/feature_extraction/count_vectorizer.ex:3]: Don't indent the docs by two spaces. They should be aligned with the closing `"""`, as that's the one that controls indentation. :)

File: lib/scholar/feature_extraction/count_vectorizer.ex [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+  This function performs:
+
+  * tokenization of input text (splitting by whitespace and removing punctuation)
+  * vocabulary construction
+  * creation of a count tensor
+
+  ## Options
+  #{NimbleOptions.docs(@binarize_schema)}
+  ## Examples
```
Comment (josevalim) [inline lib/scholar/feature_extraction/count_vectorizer.ex:39]:
```suggestion
  ## Options

  #{NimbleOptions.docs(@binarize_schema)}

  ## Examples

```

File: lib/scholar/feature_extraction/count_vectorizer.ex [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      if opts[:indexed_tensor] do
+        {corpus, %{}}
+      else
+        preprocessed_corpus = preprocess(corpus)
+        vocabulary = create_vocabulary(preprocessed_corpus)
+        tensor = create_tensor(preprocessed_corpus, vocabulary)
+        {tensor, vocabulary}
+      end
+
+    max_index = tensor |> Nx.reduce_max() |> Nx.add(1) |> Nx.to_number()
```
Comment (josevalim) [inline lib/scholar/feature_extraction/count_vectorizer.ex:82]: Under `jit` mode, you cannot convert a tensor to a number. You can try adding this test:

  fun = &Scholar.FeatureExtraction.CountVectorizer.fit_transform(&1, indexed_tensor: true)
  Nx.Defn.jit(fun).(Nx.tensor([[0, 1, 2], [1, 3, 4]])

The correct solution is to compute the default value for `max_index` inside `defn`.

**Review by josevalim — COMMENTED** [review]
The `fit_transform` function currently has an issue in that it accepts either tensors (which can be JITed) or a list of strings, which cannot. Generally speaking, our `fit_transform` functions can be jitted, so this would be breaking the tradition.

It seems passing a list of strings is implementing rather a simple tokenization algorithm? I think we should either move it elsewhere OR simply remove the functionality from here. My suggestion would be to remove it :)

File: lib/scholar/feature_extraction/count_vectorizer.ex (hunk 3 above)
Comment (ksew1) [reply lib/scholar/feature_extraction/count_vectorizer.ex:33]: The problem is that it needs to be a number in order to create a tensor with this shape. Inside `defn`, I'm afraid it is not possible to dynamically obtain a number. We can make this a required option, so we do not need to use `Nx.to_number()`.

File: lib/scholar/feature_extraction/count_vectorizer.ex (hunk 2 above)
Comment (ksew1) [reply lib/scholar/feature_extraction/count_vectorizer.ex]: Fixed 😄

File: lib/scholar/feature_extraction/count_vectorizer.ex (hunk 1 above)
Comment (ksew1) [reply lib/scholar/feature_extraction/count_vectorizer.ex]: Fixed 😄

**Comment by ksew1** [comment]
> The `fit_transform` function currently has an issue in that it accepts either tensors (which can be JITed) or a list of strings, which cannot. Generally speaking, our `fit_transform` functions can be jitted, so this would be breaking the tradition.
> 
> It seems passing a list of strings is implementing rather a simple tokenization algorithm? I think we should either move it elsewhere OR simply remove the functionality from here. My suggestion would be to remove it :)

Removed this feature, now it operates only on tensor

File: lib/scholar/feature_extraction/count_vectorizer.ex (hunk 3 above)
Comment (josevalim) [reply lib/scholar/feature_extraction/count_vectorizer.ex:33]: I see. So we should not use `deftransform`, because this certainly cannot be invoked inside a `defn`, as we don't yet support dynamic shapes. For now, it should be a regular `def` and maybe it should be called something else.

I believe this topic appeared in the past and we maybe discussed a possible contract for keeping code that doesn't work inside `defn` but I don't recall it right now. :) Maybe @msluszniak does?

Comment (msluszniak) [reply lib/scholar/feature_extraction/count_vectorizer.ex:33]: Usually, if the shape was based on computations we either droped the options that force it (like percent of variation preserved in PCA) or develop some heuristics that roughly assesses the upper bound of the problematic shape, and then make computations on bigger tensor with some "padding".

Comment (msluszniak) [reply lib/scholar/feature_extraction/count_vectorizer.ex:33]: Or make an option that is required via NimbleOptions

Comment (josevalim) [reply lib/scholar/feature_extraction/count_vectorizer.ex:33]: Maybe an option is the way to go then and we could even have a helper function like `CountVectorizer.size(n)` that they could use to compute it.

Comment (ksew1) [reply lib/scholar/feature_extraction/count_vectorizer.ex]: I made this required option and added helper function 😄

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
